python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -1553,6 +1559,290 @@ module IterableUnpacking {
+ * - keyword pattern: the appropriate attribute is read from the subject (see class pattern)
+ *
+ * Inside the class pattern, we also find positional arguments. They are converted to
+ * keyword arguments using the `__match_args__` attribute on the class. We do not
+ * currently model this.
+ */
+module MatchUnpacking {
+  /**
+   * The subject of a match flows to each top-level pattern
+   * (a pattern directly under a `case` statement).
+   *
+   * We could consider a model closer to use-use-flow, where the subject
+   * only flows to the first top-level pattern and from there to the
+   * following ones.
+   */
+  predicate matchSubjectFlowStep(Node nodeFrom, Node nodeTo) {
+    exists(MatchStmt match, Expr subject, Pattern target |
+      subject = match.getSubject() and
+      target = match.getCase(_).(Case).getPattern()
+    |
+      nodeFrom.asExpr() = subject and
+      nodeTo.asCfgNode().getNode() = target
+    )
+  }
+
+  /**
+   * as pattern: subject flows to alias as well as to the interior pattern
+   * syntax (toplevel): `case pattern as alias:`
+   */
+  predicate matchAsFlowStep(Node nodeFrom, Node nodeTo) {
+    exists(MatchAsPattern subject, Name alias | alias = subject.getAlias() |
+      nodeFrom.asCfgNode().getNode() = subject and
+      (
+        // the subject flows to the alias
+        nodeTo.asVar().getDefinition().(PatternAliasDefinition).getDefiningNode().getNode() = alias
+        or
+        // the subject flows to the interior pattern
+        nodeTo.asCfgNode().getNode() = subject.getPattern()
+      )
+    )
+  }
+
+  /**
+   * or pattern: subject flows to each alternative
+   * syntax (toplevel): `case alt1 | alt2:`
+   */
+  predicate matchOrFlowStep(Node nodeFrom, Node nodeTo) {
+    exists(MatchOrPattern subject, Pattern pattern | pattern = subject.getAPattern() |
+      nodeFrom.asCfgNode().getNode() = subject and
+      nodeTo.asCfgNode().getNode() = pattern
+    )
+  }
+
+  /**
+   * capture pattern: subject flows to the variable
+   * syntax (toplevel): `case var:`
+   */
+  predicate matchCaptureFlowStep(Node nodeFrom, Node nodeTo) {
+    exists(MatchCapturePattern capture, Name var | capture.getVariable() = var |
+      nodeFrom.asCfgNode().getNode() = capture and
+      nodeTo.asVar().getDefinition().(PatternCaptureDefinition).getDefiningNode().getNode() = var
+    )
+  }
+
+  /**
+   * sequence pattern: each element reads from subject at the associated index
+   * syntax (toplevel): `case [a, b]:`
+   */
+  predicate matchSequenceReadStep(Node nodeFrom, Content c, Node nodeTo) {
+    exists(MatchSequencePattern subject, int index, Pattern element |
+      element = subject.getPattern(index)
+    |
+      nodeFrom.asCfgNode().getNode() = subject and
+      nodeTo.asCfgNode().getNode() = element and
+      (
+        // tuple content
+        c.(TupleElementContent).getIndex() = index
+        or
+        // list content
+        c instanceof ListElementContent
```

Comment:
Do we need this always, or just when the pattern is sufficiently complicated that accurate indices are not available? 

It seems to me that for `case [a, b]:` we should know exactly which indices go where, but maybe I'm overlooking something.

Reply:
`case [a, b]` is equivalent to `case (a, b)`. Just like for utterable unpacking, they are the same internally. But `match [a, b]` is different from `match (a, b)` since they receive two different values. We also model these two values differently, as in the list case we forget about indices, just assigning the whole list `ListElementContent`. So when a case sees a sequence, it could have either list content or tuple content depending on the subject coming from above. So we have to read both.